packages/edit-site/src/components/navigation-sidebar/navigation-panel/index.js

```diff
@@ -6,35 +6,82 @@ import classnames from 'classnames';
 /**
  * WordPress dependencies
  */
+import {
+	__experimentalNavigation as Navigation,
+	__experimentalNavigationBackButton as NavigationBackButton,
+	__experimentalNavigationGroup as NavigationGroup,
+	__experimentalNavigationItem as NavigationItem,
+	__experimentalNavigationMenu as NavigationMenu,
+} from '@wordpress/components';
 import { usePrevious } from '@wordpress/compose';
 import { store as coreDataStore } from '@wordpress/core-data';
 import { useSelect, useDispatch } from '@wordpress/data';
 import { useState, useEffect, useRef } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
 import { ESCAPE } from '@wordpress/keycodes';
 
 /**
  * Internal dependencies
  */
-import ContentNavigation from './content-navigation';
-import TemplatesNavigation from './templates-navigation';
-import { MENU_ROOT } from './constants';
+import ContentPagesMenu from './menus/content-pages';
+import ContentCategoriesMenu from './menus/content-categories';
+import ContentPostsMenu from './menus/content-posts';
+import TemplatesMenu from './menus/templates';
+import TemplatePartsMenu from './menus/template-parts';
+import MainDashboardButton from '../../main-dashboard-button';
 import { store as editSiteStore } from '../../../store';
+import {
+	MENU_ROOT,
+	MENU_TEMPLATE_PARTS,
+	MENU_TEMPLATES,
+	MENU_CONTENT_CATEGORIES,
+	MENU_CONTENT_PAGES,
+	MENU_CONTENT_POSTS,
+} from './constants';
 
 const NavigationPanel = ( { isOpen } ) => {
 	const [ contentActiveMenu, setContentActiveMenu ] = useState( MENU_ROOT );
-	const { templatesActiveMenu, siteTitle } = useSelect( ( select ) => {
-		const { getNavigationPanelActiveMenu } = select( editSiteStore );
+	const {
+		page: { context: { postType, postId } = {} } = {},
+		editedPostId,
+		editedPostType,
+		templatesActiveMenu,
+		siteTitle,
+	} = useSelect( ( select ) => {
+		const {
+			getEditedPostType,
+			getEditedPostId,
+			getNavigationPanelActiveMenu,
+			getPage,
+		} = select( editSiteStore );
 		const { getEntityRecord } = select( coreDataStore );
 
 		const siteData =
 			getEntityRecord( 'root', '__unstableBase', undefined ) || {};
 
 		return {
+			page: getPage(),
+			editedPostId: getEditedPostId(),
+			editedPostType: getEditedPostType(),
 			templatesActiveMenu: getNavigationPanelActiveMenu(),
 			siteTitle: siteData.name,
 		};
 	}, [] );
 
+	const { setNavigationPanelActiveMenu } = useDispatch( editSiteStore );
+
+	const isTemplateActiveRoot = templatesActiveMenu === MENU_ROOT;
+	const isContentActiveRoot = contentActiveMenu === MENU_ROOT;
+	let activeMenu;
+	let activeItem;
+	if ( isTemplateActiveRoot && isContentActiveRoot ) {
+		activeMenu = MENU_ROOT;
+	} else {
+		[ activeMenu, activeItem ] = isTemplateActiveRoot
+			? [ contentActiveMenu, `${ postType }-${ postId }` ]
+			: [ templatesActiveMenu, `${ editedPostType }-${ editedPostId }` ];
+	}
```

Comment:
Although I love one-liners, I think this would easier to understand if we had simple `if` branches.

```suggestion
	if ( isTemplateActiveRoot && isContentActiveRoot ) {
		activeMenu = MENU_ROOT;
	} else if ( isTemplateActiveRoot) {
		activeMenu = contentActiveMenu
		activeItem = `${ postType }-${ postId }`
	} else {
	        activeMenu = templatesActiveMenu
		activeItem = `${ editedPostType }-${ editedPostId }`
	}
```